stdlib/public/Synchronization/Mutex/LinuxImpl.swift

```diff
@@ -0,0 +1,137 @@
+//
+// This source file is part of the Swift Atomics open source project
+//
+// Copyright (c) 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SynchronizationShims
+
+@_alwaysEmitIntoClient
+@_extern(c, "gettid")
+func _gettid() -> UInt32
+
+extension Atomic where Value == UInt32 {
+  borrowing func wait() {
+    _swift_stdlib_wait(.init(rawAddress))
+  }
+
+  borrowing func wake() {
+    _swift_stdlib_wake(.init(rawAddress))
+  }
+}
+
+@available(SwiftStdlib 5.11, *)
+@frozen
+@usableFromInline
+@_staticExclusiveOnly
+internal struct _MutexHandle: ~Copyable {
+  // There are only 3 different values that storage can hold at a single time.
+  // 0: unlocked
+  // _gettid: locked, current thread's id (uncontended)
+  // (_gettid | SWIFT_FUTEX_WAITERS): locked, current thread's id (contended)
+  @usableFromInline
+  let storage: Atomic<UInt32>
+
+  @available(SwiftStdlib 5.11, *)
+  @_alwaysEmitIntoClient
+  @_transparent
+  init() {
+    storage = Atomic(0)
+  }
+
+  @available(SwiftStdlib 5.11, *)
+  @usableFromInline
+  borrowing func lock() {
+    // TODO: Is it worth caching this value in TLS?
+    let selfId = _gettid()
+
+    // Note: We could probably merge this cas into a do/while style loop, but we
+    // really want to perform the strong variant before attempting to do weak
+    // ones in the loop.
+
+    var (exchanged, state) = storage.compareExchange(
+      expected: 0,
+      desired: selfId,
+      successOrdering: .acquiring,
+      failureOrdering: .relaxed
+    )
+
+    if _fastPath(exchanged) {
+      // Locked!
+      return
+    }
+
+    while !exchanged {
+      // Clear the waiter bit, if we have one, and check to see if this is a
+      // recursive call to lock.
+      let currentOwner = state & ~SWIFT_FUTEX_WAITERS
+      if _slowPath(currentOwner == selfId) {
+        // TODO: Need a colder function.
+        fatalError("Recursive call to lock")
+      }
+
+      // Block until unlock has been called. This will return early if the call
+      // to unlock happened between attempting to acquire and attempting to
+      // wait while nobody else managed to acquire it yet.
+      storage.wait()
+
+      (exchanged, state) = storage.weakCompareExchange(
+        expected: 0,
+        desired: selfId,
+        successOrdering: .acquiring,
+        failureOrdering: .relaxed
+      )
+    }
+
+    // Locked!
+  }
+
+  @available(SwiftStdlib 5.11, *)
+  @usableFromInline
+  borrowing func tryLock() -> Bool {
+    storage.compareExchange(
+      expected: 0,
+      desired: _gettid(),
+      successOrdering: .acquiring,
+      failureOrdering: .relaxed
+    ).exchanged
```

Comment:
Should we use `FUTEX_TRYLOCK_PI` here? There are situations where this compare-exchange will fail but `FUTEX_TRYLOCK_PI` would succeed.

Reply:
Posting from offline chat:

Alejandro:
> so we could call FUTEX_TRYLOCK_PI, but at least according to the man page it’s useful to call this when the value gets stale because the acquired thread might’ve died. at least in Swift currently, threads that die sort of bring the entire process down with it and we’re generally not great at recovering in such cases. I’m unsure if there are other ways the waiter bit could go stale 🤔 

Alastair:
> The TRYLOCK call is clearly optional; I think everything will work without it, though there’s a risk we might get into a situation where a lock can’t ever be taken with the try-lock routine (but I suspect you’re right that it’s unlikely to happen in a Swift program).

Reply:
All that said, I was thinking, if we can detect the waiter bit being set, we could call `FUTEX_TRYLOCK_PI` in that case, which might avoid any problems in future?